### ISSUE TITLE
Request: update domain boundary potentials during simulation execution

### ISSUE THREAD

**Issue by HuntFeng** [ECP-WarpX/WarpX#4715]
To reproduce the error, build WarpX in the following way

```
warpx=WarpX-24.02
echo "compile $warpx"
cmake -S $HOME/$warpx -B $HOME/$warpx/build -DWarpX_DIMS=RZ \
  -DWarpX_COMPUTE=CUDA \
  -DWARX_MPI=ON \
  -DWarpX_QED=OFF \
  -DWarpX_OPENPMD=ON \
  -DWarpX_PYTHON=ON

echo "build warpx and do pip install"
cmake --build $HOME/$warpx/build --target pip_install -j 16
```

Then use the example input [spacecraft_charging/PICMI_inputs_rz.py](https://github.com/ECP-WarpX/WarpX/blob/development/Examples/Physics_applications/spacecraft_charging/PICMI_inputs_rz.py).

Here is the output file and Backtrace file
[output.log](https://github.com/ECP-WarpX/WarpX/files/14375623/output.log)
[Backtrace.0.txt](https://github.com/ECP-WarpX/WarpX/files/14375628/Backtrace.0.txt)

I also tried my own input script, it runs fine as long as we don't access fields 
```
phi_wrapper = fields.PhiFPWrapper()
phi_wrapper[Ellipsis] # Segfault occurs when accessing fields data
```

Edit:

According the discussion, it seems there is no simple way to access fields data in GPU memory. I wonder if WarpX can implement this feature to read/write fields data in GPU?

**Comment by roelof-groenewald** [comment]
I expect the error might occur because the field data is located in device (GPU) memory. Try running with managed memory turned on (pass `warpx_amrex_the_arena_is_managed=True` to the simulation object) to allow WarpX to automatically copy data over to the CPU. Or you can install `cupy` to have it handle the device-to-host copy.

**Comment by HuntFeng** [comment]
> I expect the error might occur because the field data is located in device (GPU) memory. Try running with managed memory turned on (pass `warpx_amrex_the_arena_is_managed=True` to the simulation object) to allow WarpX to automatically copy data over to the CPU. Or you can install `cupy` to have it handle the device-to-host copy. 

Thank you for the suggestion. 

Is there a way to access GPU memory directly without copying the data to CPU? I'm asking this because it seems possible to do so with particle_container data.

**Comment by roelof-groenewald** [comment]
Not via the `fields.py` wrapper's `__getitem__` function, no. That function collects the field data from all processes and gathers it to all processes while copying it to CPU. You can access the field data directly (on GPU) by iterating over a field multifab. I don't think there is an example yet of how to do this.

**Comment by HuntFeng** [comment]
> Not via the `fields.py` wrapper's `__getitem__` function, no. That function collects the field data from all processes and gathers it to all processes while copying it to CPU. You can access the field data directly (on GPU) by iterating over a field multifab. I don't think there is an example yet of how to do this.

Thanks for the info. I will let this issue stay open for a while and see if WarpX team would like to implement a getter/setter for field data in GPU memory.

**Comment by roelof-groenewald** [comment]
Just to clarify, you can get and set field data directly on the GPU. It would look something like:
```
phi_wrapper = fields.PhiFPWrapper()
for mfi in phi_wrapper.mf:
    device_arr = self._get_field(mfi)
    # use device_arr for something
```
but generally you need field data over the whole simulation domain or a region of the domain not necessarily enclosed inside a single tile of the underlying multifab, so you generally need to gather sections of the fields from different GPUs onto a single memory block which is where the copy comes from.
In other words, the functionality exists to access fields directly on GPU but only in special circumstances will that be helpful.

We can discuss whether a specific use case you have in mind can benefit from such an approach, otherwise we can probably close this issue.

**Comment by HuntFeng** [comment]
Here is my use case:

I want to set boundary values using custom algorithms. Therefore I need to read/write the field data each time step.

We can already do this in CPU simulations through the use of MultiFabWrappers in fields.py. I just need the same functionality in GPU simulation.

**Comment by roelof-groenewald** [comment]
Have you run it with the `warpx_amrex_the_arena_is_managed` parameter set to true? The exact same input script from CPU should run fine on GPU then and I don't expect it will make the simulation much slower than optimal.

**Comment by HuntFeng** [comment]
Yes, I set `warpx_amrex_the_arena_is_managed=True` and it works well. 

Since you mentioned the data will be copied to CPU using this method, I was wondering if we can directly modify data on GPU to avoid the data transferring between GPU and CPU. In my knowledge the data transferring between GPU and CPU is often time-consuming, I don't know if this will make the simulation slow.

**Comment by HuntFeng** [comment]
Update:

Using warpx_amrex_the_arena_is_managed=True I can access the field datas without issues, but I am unable to set values to the fields.

Like I mentioned before, I am using a custom algorithm to determine the boundary values of the fields. Is this possible with GPU backend?

**Comment by roelof-groenewald** [comment]
We figured out that the actual issue here is that currently the domain boundary potentials cannot be updated during a simulation. The PR linked above will fix that. I'll change the issue title accordingly.

@HuntFeng could you please try out the changes on the linked PR and let me know if it works for your purpose? After building from the PR branch you should be able to update the domain boundary potential using:
```
sim.extension.warpx.set_potential_on_domain_boundary(potential_zhi=f"{new_potential}")
```
similarly to how @EyaDammak did it in her spacecraft charging example.

**Comment by HuntFeng** [comment]
Thank you @roelof-groenewald for your hard work!

**Comment by roelof-groenewald** [comment]
> Thank you @roelof-groenewald for your hard work!

No problem. Please let us know once you've had a chance to test out the new changes, and we can merge it into `development`.
